### PR TITLE
Implement zero-copy within UDPNetProcessorInternal::udp_read_from_net

### PR DESCRIPTION
**Pull request by oknet**

### CODE REVIEW

**Comment by oknet** [comment]
[approve ci clang-format]

**Comment by oknet** [comment]
[approve ci autest]

**Comment by oknet** [comment]
And I have an idea that try to detect the udp packet size first.

```
diff --git a/iocore/net/UnixUDPNet.cc b/iocore/net/UnixUDPNet.cc
index 2143cf6..8036f75 100644
--- a/iocore/net/UnixUDPNet.cc
+++ b/iocore/net/UnixUDPNet.cc
@@ -121,6 +121,24 @@ UDPNetProcessorInternal::udp_read_from_net(UDPNetHandler *nh, UDPConnection *xuc
     unsigned niov = 0;
     IOBufferBlock *b, *last;
 
+#ifdef DETECT_UDPPACKET_SIZE
+    // detect the udp pakcet size
+    struct iovec iov_detect;
+    iov_detect.iov_base = nullptr;
+    iov_detect.iov_len = 0;
+    struct msghdr detect_msg = {0};
+    detect_msg.msg_iov = &iov_detect;
+    detect_msg.msg_iovlen = 1;
+    int detect_msg_size = socketManager.recvmsg(uc->getFd(), &detect_msg, MSG_PEEK | MSG_TRUNC);
+    if (detect_msg_size < 0) {
+      break;
+    }
+    max_niov = (detect_msg_size / buffer_size);
+    if (detect_msg_size % buffer_size) {
+      max_niov++;
+    }
+#endif
+
     // build struct iov
     // reuse the block in chain if available
     b    = chain.get();
```

Any suggestion ? Is it a good idea ?

**Comment by maskit** [comment]
So, it uses `IOBufferBlock::clone` instead of `memcpy` to avoid copy, right? When will the original blocks be freed or reused? The both buffers (original and cloned) would have the same life time, and the original buffer shouldn't be freed until applications process the packet.

**Comment by oknet** [comment]
@maskit Changes list below from your last review:

```
diff --git a/iocore/net/P_UDPPacket.h b/iocore/net/P_UDPPacket.h
index 0c37a39..154db9f 100644
--- a/iocore/net/P_UDPPacket.h
+++ b/iocore/net/P_UDPPacket.h
@@ -207,7 +207,7 @@ new_UDPPacket(struct sockaddr const *to, ink_hrtime when, IOBufferBlock *buf, in
 }
 
 TS_INLINE UDPPacket *
-new_UDPPacket(struct sockaddr const *to, ink_hrtime when, Ptr<IOBufferBlock> buf)
+new_UDPPacket(struct sockaddr const *to, ink_hrtime when, Ptr<IOBufferBlock> &buf)
 {
   UDPPacketInternal *p = udpPacketAllocator.alloc();
 
@@ -246,6 +246,41 @@ new_incoming_UDPPacket(struct sockaddr *from, Ptr<IOBufferBlock> &buf)
 }
 
 TS_INLINE UDPPacket *
+new_incoming_UDPPacket(struct sockaddr *from, Ptr<IOBufferBlock> &buf)
+{
+  UDPPacketInternal *p = udpPacketAllocator.alloc();
+
+  p->in_the_priority_queue = 0;
+  p->in_heap               = 0;
+  p->delivery_time         = 0;
+  ats_ip_copy(&p->from, from);
+  p->chain = buf;
+
+  return p;
+}
+
+TS_INLINE UDPPacket *
 new_UDPPacket()
 {
   UDPPacketInternal *p = udpPacketAllocator.alloc();
```

and the comment to explain why is 65536.
```
+  // The max length of receive buffer is NET_MAX_IOV (16) * buffer_size (4096) = 65536 bytes.
+  // Because the 'UDP Length' is type of uint16_t defined in RFC 768.
+  // And there is 8 octets in 'User Datagram Header' which means the max length of payload is no more than 65527 bytes.
```

and change to call `new_incoming_UDPPacket(struct sockaddr *from, Ptr<IOBufferBlock> &buf)`:
```
     // create packet
-    UDPPacket *p = new_incoming_UDPPacket(ats_ip_sa_cast(&fromaddr), buf, r);
+    UDPPacket *p = new_incoming_UDPPacket(ats_ip_sa_cast(&fromaddr), chain);
```

**Comment by oknet** [comment]
@maskit @masaori335 

1. replace `buf[65536]` with a IOBufferBlock chain
- 4KB / block and 16 blocks.
2. replace `recvfrom` with `recvmsg`
- the `recvmsg` is `readv` of `read`.
- it saved data into multiple blocks in the chain.
3. create `UDPPacket` by the chain.
- it increase the refcount for the chain rather than copy the data by `memcpy`.

**Comment by maskit** [comment]
I haven't looked into the details but a couple of comments for now.

- IPv6
UDP packet length can be over 64K on IPv6 Jumboframes.
https://tools.ietf.org/html/rfc2675#section-4

- `recvmmsg`
Because this changes how we receive UDP packets, we should consider use of `recvmmsg`.

**Comment by oknet** [comment]
@maskit 

Is the `Jumboframes` is standard support for ATS ?

The `recvmmsg()` is Linux-specific.

**Comment by oknet** [comment]
Rebased on master since PR #2716 merged

**Comment by maskit** [comment]
I don't think we should support them right now, but we should think proactively. I wanted to confirm that the design is extensible and won't prevent using them.

The branch has a merge commit (is not rebased).

**Comment by masaori335** [comment]
FYI, I created a issue for `recvmmsg()` / `sendmmsg()` support.

File: iocore/net/P_UDPPacket.h [hunk 1]
```diff
@@ -246,6 +246,41 @@ new_incoming_UDPPacket(struct sockaddr *from, char *buf, int len)
 }
 
 TS_INLINE UDPPacket *
+new_incoming_UDPPacket(struct sockaddr *from, IOBufferBlock *block, int len)
+{
+  (void)len;
```
Comment (masaori335) [inline iocore/net/P_UDPPacket.h]: Why don't you comment out `len` ?

Comment (masaori335) [reply iocore/net/P_UDPPacket.h]: or just remove it like `new_incoming_UDPPacket(struct sockaddr *from, Ptr<IOBufferBlock> &block);`

File: iocore/net/P_UDPPacket.h [hunk 2]
```diff
@@ -246,6 +246,41 @@ new_incoming_UDPPacket(struct sockaddr *from, char *buf, int len)
...
+    body = block->clone();
+    p->append_block(body);
+    block = block->next.get();
+  }
+
+  return p;
+}
+
+TS_INLINE UDPPacket *
+new_incoming_UDPPacket(struct sockaddr *from, Ptr<IOBufferBlock> &buf)
```
Comment (masaori335) [inline iocore/net/P_UDPPacket.h]: In the prototype of this function, this `buf` is called `block`. It'd be great if you pick out one of them for consistency :)

File: iocore/net/P_UDPPacket.h [hunk 3]
```diff
@@ -246,6 +246,41 @@ new_incoming_UDPPacket(struct sockaddr *from, char *buf, int len)
 }
 
 TS_INLINE UDPPacket *
+new_incoming_UDPPacket(struct sockaddr *from, IOBufferBlock *block, int len)
```
Comment (maskit) [inline iocore/net/P_UDPPacket.h]: It seems like this one and the original one are not used anymore. Should we keep them?

**Comment by maskit** [comment]
> And I have an idea that try to detect the udp packet size first.

Maybe the logic would be needed when we support >64KB datagrams. For now, it would be better to print an error message if a datagram did not fit in the buffer.

File: iocore/net/P_UDPPacket.h (hunk 2 above)
Comment (oknet) [reply iocore/net/P_UDPPacket.h]: pick `block` as the name.

File: iocore/net/P_UDPPacket.h (hunk 1 above)
Comment (oknet) [reply iocore/net/P_UDPPacket.h]: remove it.

File: iocore/net/P_UDPPacket.h (hunk 3 above)
Comment (oknet) [reply iocore/net/P_UDPPacket.h]: remove it.

**Comment by oknet** [comment]
@maskit

Now, it is detect packet size first and then read the packet.
Throw a debug message if Huge Packet received or packet truncated.

**Comment by maskit** [comment]
I looked into `recvmmsg()` deeply and realized that the size detection wouldn't work with it. I have no idea how we can support jumbograms.  I think checking MSG_TRUNC is enough for now.

https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/net/socket.c#n2272

**Comment by oknet** [comment]
The udp packet size detect code is removed.

It is can be revert by the below patch:

```
diff --git a/iocore/net/UnixUDPNet.cc b/iocore/net/UnixUDPNet.cc
index 4aae118..dab24d6 100644
--- a/iocore/net/UnixUDPNet.cc
+++ b/iocore/net/UnixUDPNet.cc
@@ -126,27 +126,6 @@ UDPNetProcessorInternal::udp_read_from_net(UDPNetHandler *nh, UDPConnection *xuc
     unsigned int niov;
     IOBufferBlock *b, *last;
 
-    // detect the udp pakcet size first
-    struct iovec iov_detect;
-    iov_detect.iov_base = nullptr;
-    iov_detect.iov_len = 0;
-    ink_zero(msg);
-    msg.msg_iov = &iov_detect;
-    msg.msg_iovlen = 1;
-    int64_t msg_size = socketManager.recvmsg(uc->getFd(), &msg, MSG_PEEK | MSG_TRUNC);
-    if (msg_size < 0) {
-      break;
-    }
-    max_niov = (msg_size / buffer_size);
-    if (msg_size % buffer_size) {
-      max_niov++;
-    }
-    if (max_niov > NET_MAX_IOV) {
-      Debug("udp-read", "Huge UDP packet size = %" PRId64 "", msg_size);
-      max_niov = NET_MAX_IOV;
-    }
-    // detect end
-
     // build struct iov
     // reuse the block in chain if available
     b    = chain.get();
@@ -186,8 +165,8 @@ UDPNetProcessorInternal::udp_read_from_net(UDPNetHandler *nh, UDPConnection *xuc
     }
 
     // truncated check
-    if (msg.msg_flags & MSG_TRUNC || r < msg_size) {
-      Debug("udp-read", "The UDP packet is truncated, full size = %" PRId64 ", reads = %" PRId64 "", msg_size, r);
+    if (msg.msg_flags & MSG_TRUNC) {
+      Debug("udp-read", "The UDP packet is truncated");
     }
 
     // fill the IOBufferBlock chain
```

**Pull request merged**